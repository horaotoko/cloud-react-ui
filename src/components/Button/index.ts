export { ButtonContainer as Button } from "./ButtonContainer";
export type { ButtonProps } from "./ButtonContainer";
