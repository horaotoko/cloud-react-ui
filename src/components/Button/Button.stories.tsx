import { text, radios, boolean } from "@storybook/addon-knobs";
import { ButtonContainer, ButtonProps } from "./ButtonContainer";

export default {
  component: ButtonContainer,
  title: "components/Button",
};

const sizeOptions: { [key: string]: ButtonProps["size"] } = {
  small: "small",
  medium: "medium",
  large: "large",
};

const colorOptions: { [key: string]: ButtonProps["color"] } = {
  default: "default",
  danger: "danger",
  primary: "primary",
  settings: "settings",
};

const iconOptions: { [key: string]: ButtonProps["icon"] } = {
  none: undefined,
  plus: "plus",
};

const iconPlacementOptions: { [key: string]: ButtonProps["iconPlacement"] } = {
  start: "start",
  end: "end",
};

export const Button: React.FC = () => (
  <ButtonContainer
    size={radios("Size", sizeOptions, "small") as ButtonProps["size"]}
    color={radios("Color", colorOptions, "default") as ButtonProps["color"]}
    disabled={boolean("Disabled", false)}
    icon={radios("Icon", iconOptions, undefined) as ButtonProps["icon"]}
    iconPlacement={
      radios(
        "IconPlacement",
        iconPlacementOptions,
        undefined
      ) as ButtonProps["iconPlacement"]
    }
    onClick={() => console.log("button clicked")}
  >
    {text("Button Text", "Button")}
  </ButtonContainer>
);
