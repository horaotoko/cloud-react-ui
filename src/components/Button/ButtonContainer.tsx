import { ButtonPresenter } from "./ButtonPresenter";
import { IconTypes } from "../Icon/IconContainer";

export type ButtonProps = {
  size?: "small" | "medium" | "large";
  color?: "default" | "danger" | "primary" | "settings";
  disabled?: boolean;
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"];
  icon?: IconTypes;
  iconPlacement?: "start" | "end";
  children?: React.ReactNode;
  className?: string;
};

export function ButtonContainer(props: ButtonProps): React.ReactElement {
  return <ButtonPresenter {...props} />;
}
