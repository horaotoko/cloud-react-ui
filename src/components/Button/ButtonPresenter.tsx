import React from "react";
import styled, { css } from "styled-components";
import { ButtonProps } from "./ButtonContainer";
import { Icon } from "../Icon";
import { defaultProps } from "../../theme";

const StyledButton = styled.button<ButtonProps>`
  ${({ theme, size = "medium", color = "default", disabled }) => {
    const buttonSize = theme.button.size[size];
    const buttonColor = theme.button[disabled ? "disabled" : color];

    return css`
      && {
        display: inline-block;
        height: ${buttonSize.height};
        padding: ${buttonSize.padding};
        line-height: ${buttonSize.height};
        font-size: ${buttonSize.fontSize};
        border-radius: ${theme.button.borderRadius};
        border: ${buttonColor.border};
        color: ${buttonColor.textColor};
        background: transparent;
        cursor: ${disabled ? "not-allowed" : "pointer"};
        position: relative;
        z-index: 0;
      }

      &&:before,
      &&:after {
        content: "";
        width: 100%;
        height: 100%;
        position: absolute;
        left: 0;
        border-radius: ${theme.button.borderRadius};
      }
      &&:before {
        z-index: -1;
        background: ${buttonColor.background};
        transition: opacity 0.2s ease-in-out;
      }
      &&:after {
        content: "";
        z-index: -2;
        background: ${buttonColor.hover.background};
        border-radius: ${theme.button.borderRadius};
      }
      &&:hover {
        &:before {
          opacity: 0;
        }
      }
      &&:active,
      &&:focus,
      &&:visited {
        outline: 0;
        box-shadow: ${buttonColor.active.boxShadow};
      }
    `;
  }}
`;
StyledButton.defaultProps = defaultProps;

const StyledIcon = styled((props) => <Icon {...props} />)`
  ${({ theme, iconPlacement = "start", color = "default", disabled }) => {
    return css`
      && {
        padding-${iconPlacement === "start" ? "right" : "left"}: 4px;
        width: 10px;
        color: ${theme.button[disabled ? "disabled" : color].iconColor};
      }
    `;
  }}
`;
StyledIcon.defaultProps = defaultProps;

export function ButtonPresenter({
  size,
  type,
  icon,
  iconPlacement,
  children,
  ...rest
}: ButtonProps): React.ReactElement {
  const ButtonIcon = (
    <StyledIcon icon={icon} iconPlacement={iconPlacement} {...rest} />
  );

  return (
    <StyledButton size={size} type={type} {...rest}>
      {icon && iconPlacement === "start" && ButtonIcon}
      {children}
      {icon && iconPlacement === "end" && ButtonIcon}
    </StyledButton>
  );
}
