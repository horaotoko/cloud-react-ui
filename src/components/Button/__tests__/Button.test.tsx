import { render, fireEvent } from "@testing-library/react";
import { Button } from "../Button";

describe("Button", () => {
  it("default", () => {
    const { asFragment } = render(<Button>button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("with icon", () => {
    const { asFragment } = render(
      <Button icon="bell" iconPlacement="start">
        button
      </Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
