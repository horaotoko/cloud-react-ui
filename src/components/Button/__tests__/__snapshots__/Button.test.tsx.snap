// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default 1`] = `
<DocumentFragment>
  .c0.c0 {
  display: inline-block;
  height: 32px;
  padding: 0 12px;
  line-height: 32px;
  font-size: 14px;
  border-radius: 4px;
  border: 1px solid #D4D8DD;
  color: #333333;
  background: transparent;
  cursor: pointer;
  position: relative;
  z-index: 0;
}

.c0.c0:before,
.c0.c0:after {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  border-radius: 4px;
}

.c0.c0:before {
  z-index: -1;
  background: linear-gradient(to bottom,#FFFFFF,#ECF2FD);
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
}

.c0.c0:after {
  content: '';
  z-index: -2;
  background: linear-gradient(to bottom,#ECF2FD,#FFFFFF);
  border-radius: 4px;
}

.c0.c0:hover:before {
  opacity: 0;
}

.c0.c0:active,
.c0.c0:focus,
.c0.c0:visited {
  outline: 0;
  box-shadow: 0 0 0 2px rgba(212,216,221,0.3);
}

<button
    class="c0"
  >
    button
  </button>
</DocumentFragment>
`;

exports[`Button with icon 1`] = `
<DocumentFragment>
  .c1 {
  font-size: 13px;
  color: #333333;
}

.c1:hover {
  color: #333333;
}

.c0.c0 {
  display: inline-block;
  height: 32px;
  padding: 0 12px;
  line-height: 32px;
  font-size: 14px;
  border-radius: 4px;
  border: 1px solid #D4D8DD;
  color: #333333;
  background: transparent;
  cursor: pointer;
  position: relative;
  z-index: 0;
}

.c0.c0:before,
.c0.c0:after {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  left: 0;
  border-radius: 4px;
}

.c0.c0:before {
  z-index: -1;
  background: linear-gradient(to bottom,#FFFFFF,#ECF2FD);
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
}

.c0.c0:after {
  content: '';
  z-index: -2;
  background: linear-gradient(to bottom,#ECF2FD,#FFFFFF);
  border-radius: 4px;
}

.c0.c0:hover:before {
  opacity: 0;
}

.c0.c0:active,
.c0.c0:focus,
.c0.c0:visited {
  outline: 0;
  box-shadow: 0 0 0 2px rgba(212,216,221,0.3);
}

.c2.c2 {
  padding-right: 4px;
  width: 10px;
  color: #BEBEBE;
}

<button
    class="c0"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-bell fa-w-14 c1 c2"
      data-icon="bell"
      data-prefix="fas"
      focusable="false"
      role="img"
      theme="[object Object]"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 512c35.32 0 63.97-28.65 63.97-64H160.03c0 35.35 28.65 64 63.97 64zm215.39-149.71c-19.32-20.76-55.47-51.99-55.47-154.29 0-77.7-54.48-139.9-127.94-155.16V32c0-17.67-14.32-32-31.98-32s-31.98 14.33-31.98 32v20.84C118.56 68.1 64.08 130.3 64.08 208c0 102.3-36.15 133.53-55.47 154.29-6 6.45-8.66 14.16-8.61 21.71.11 16.4 12.98 32 32.1 32h383.8c19.12 0 32-15.6 32.1-32 .05-7.55-2.61-15.27-8.61-21.71z"
        fill="currentColor"
      />
    </svg>
    button
  </button>
</DocumentFragment>
`;
